worker/peergrouper/desired.go

```diff
@@ -397,88 +399,128 @@ func (p *peerGroupChanges) getMachinesVoting() {
 // If the check is satisfied, then a warning is logged and no change is made.
 // Otherwise an error is returned to indicate that a HA space must be
-// configured in order to proceed.
-func (p *peerGroupChanges) updateAddressFromInternal(id string, info *peerGroupInfo) error {
-	member := p.members[id]
-	hostPorts := info.machines[id].GetPotentialMongoHostPorts(info.mongoPort)
-	addrs := network.SelectInternalHostPorts(hostPorts, false)
-
-	// This should not happen because SelectInternalHostPorts will choose a
-	// public address when there are no cloud-local addresses.
-	// Zero addresses would mean the machine would not be accessible at all.
-	// We ignore this outcome and leave the address alone.
-	if len(addrs) == 0 {
-		return nil
-	}
+// configured in order to proceed. Such machines have their status set to
+// indicate that they require intervention.
+func (p *peerGroupChanges) updateAddressesFromInternal(info *peerGroupInfo) error {
+	var multipleAddresses []string
 
-	if len(addrs) == 1 {
-		addr := addrs[0]
-		logger.Debugf("machine %q selected address %q by scope from %v", id, addr, hostPorts)
+	for id := range p.members {
+		m := info.machines[id]
+		hostPorts := m.GetPotentialMongoHostPorts(info.mongoPort)
+		addrs := network.SelectInternalHostPorts(hostPorts, false)
+
+		// This should not happen because SelectInternalHostPorts will choose a
+		// public address when there are no cloud-local addresses.
+		// Zero addresses would mean the machine is completely inaccessible.
+		// We ignore this outcome and leave the address alone.
+		if len(addrs) == 0 {
+			continue
+		}
 
-		if member.Address != addr {
-			member.Address = addr
-			p.isChanged = true
+		// Unique address; we can use this for Mongo peer communication.
+		member := p.members[id]
+		if len(addrs) == 1 {
+			addr := addrs[0]
+			logger.Debugf("machine %q selected address %q by scope from %v", id, addr, hostPorts)
+
+			if member.Address != addr {
+				member.Address = addr
+				p.isChanged = true
+			}
+			continue
+		}
+
+		// Multiple potential Mongo addresses.
+		// Checks are required in order to use it as a peer.
+		unchanged := false
+		if _, ok := info.recognised[id]; ok {
+			for _, addr := range addrs {
+				if member.Address == addr {
+					logger.Warningf("%s\npreserving member with unchanged address %q", multiAddressMessage, addr)
+					unchanged = true
+					break
+				}
+			}
 		}
-		return nil
-	}
 
-	msg := fmt.Sprintf(
-		"machine %q has more than one usable address and juju-ha-space is not set"+
-			"\nrun \"juju config juju-ha-space=<name>\" to set a space for Mongo peer communication", id)
+		// If this member was not previously in the replica-set, or if its
+		// address has changed, we enforce the policy of requiring a
+		// configured HA space when there are multiple cloud-local addresses.
+		if !unchanged {
+			multipleAddresses = append(multipleAddresses, id)
+			if err := m.stm.SetStatus(getStatusInfo(multiAddressMessage)); err != nil {
```

Comment:
I don't want the status message to get unwieldy, but I'm wondering if we should be including some context about what addresses might want to be used, and why that's a problem.
Maybe that should be a log message rather than a status?

Reply:
I think the status message is probably the right place given that the log message can now represent issues with multiple machines.

Reply:
I just realized a problem with setting Blocked, is that we also need to be setting status back to normal if we've ever been blocked, otherwise it would stay broken even once you've fixed it.
I like highlighting (somehow) that something is wrong that won't fix itself without human intervention.

Reply:
Ah, yes. I will address this.

Regarding the status itself. Re-reading the code in _status/status.go_, it seems that a status of _error_ might be best - it is a intended to imply the need for manual intervention and is listed as a machine state, where _blocked_ is not.

Reply:
What I overlooked is that a MachineTracker is not created unless the machine status is _started_. I am not yet sure how to go about sorting this out.

Reply:
FWIW, with my "juju remove-machine" patch, I went ahead and changed this to be:
Status != Pending

Because in that patch, it is possible for an active controller to go into Stopped (the Machiner notices that the agent is Dying, so flags it as Stopped once it has seen that case). But the agent is still actually running, and needs to be tracked.
That said, I'm still ok going with Started with our messages rather than Error.